src/server/acl/acl_family.cc

```diff
@@ -161,12 +184,38 @@ std::variant<User::UpdateRequest, ErrorReply> ParseAclSetUser(CmdArgList args) {
   return req;
 }
 
+using MaterializedContents = std::optional<std::vector<std::vector<std::string_view>>>;
+
+MaterializedContents MaterializeFileContents(std::vector<std::string>* usernames,
+                                             std::string_view file_contents) {
+  // This is fine, a very large file will top at 1-2 mb. And that's for 5000+ users with 400
+  // characters of ACL infor...
+  std::vector<std::string_view> commands = absl::StrSplit(file_contents, "\n");
+  std::vector<std::vector<std::string_view>> materialized;
+  materialized.reserve(commands.size());
+  usernames->reserve(commands.size());
+  for (auto& command : commands) {
+    if (command.empty())
+      continue;
+    std::vector<std::string_view> cmds = absl::StrSplit(command, ' ');
+    if (cmds[0] != "ACL" || cmds[1] != "SETUSER" || cmds.size() < 3) {
+      return {};
+    }
+    usernames->push_back(std::string(cmds[2]));
+    cmds.erase(cmds.begin(), cmds.begin() + 3);
+    materialized.push_back(cmds);
+  }
+  return materialized;
+}
+
 }  // namespace
 
-void AclFamily::StreamUpdatesToAllProactorConnections(std::string_view user, uint32_t update_cat) {
-  auto update_cb = [user, update_cat]([[maybe_unused]] size_t id, util::Connection* conn) {
+void AclFamily::StreamUpdatesToAllProactorConnections(const std::vector<std::string>& user,
+                                                      const std::vector<uint32_t>& update_cat) {
```

Comment:
This is called for load and acl changes that occur rarely right? If it ever becomes a bottleneck you can pass the message data inside a shared_ptr so you don't have to copy the vectors every time

Reply:
and single changes are now propagated with a vector...

Reply:
Single changes are indeed a vector with a single element, this is insignificant....

> This is called for load and acl changes that occur rarely right? If it ever becomes a bottleneck you can pass the message data inside a shared_ptr so you don't have to copy the vectors every time

Yes but I don't think it will so I opted in for the simplest change!